### PR TITLE
Mustache array empty

### PR DESCRIPTION
**Pull request by mingwli**
[AB#188556](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/188556)

For tracking [Stubble#92](https://github.com/StubbleOrg/Stubble/issues/92)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5730)

### CODE REVIEW

File: src/docfx/template/MustacheTemplate.cs [hunk 1]
```diff
@@ -44,13 +47,16 @@ private static RendererSettingsBuilder UseJson(RendererSettingsBuilder settings,
         {
             // JObject implements IEnumerable, stubble treats IEnumerable as array,
             // need to put it to section blacklist and overwrite the truthy check method.
-            var sectionBlacklist = RendererSettingsDefaults.DefaultSectionBlacklistTypes();
-            sectionBlacklist.Add(typeof(JObject));
-
             return settings.AddValueGetter(typeof(JObject), GetJObjectValue)
+                           .AddValueGetter(typeof(JArray), GetJArrayValue)
+                           .AddValueGetter(typeof(IList), OverwriteValuleGetter) // overwrite default value getters in stubble
```
Comment (mingwli) [inline src/docfx/template/MustacheTemplate.cs:52]: refer to https://github.com/StubbleOrg/Stubble/blob/master/src/Stubble.Core/Settings/RendererSettingsDefaults.cs

File: src/docfx/template/MustacheTemplate.cs [hunk 2]
```diff
@@ -64,6 +70,25 @@ private static RendererSettingsBuilder UseJson(RendererSettingsBuilder settings,
...
+                    default:
+                        return int.TryParse(key, out var index) && index > -1 && jArr.Count > index
+                            ? jArr[index] is JValue scalar
+                                ? scalar.Value ?? JValue.CreateNull()
+                                : jArr[index]
+                            : JValue.CreateNull();
+                }
+            }
+
+            object? OverwriteValuleGetter(object value, string key, bool ignoreCase) => null;
```
Comment (herohua) [inline src/docfx/template/MustacheTemplate.cs:91]: typo: `Valule` -> `Value`

File: src/docfx/template/MustacheTemplate.cs [hunk 3]
```diff
@@ -64,6 +70,25 @@ private static RendererSettingsBuilder UseJson(RendererSettingsBuilder settings,
 
                 return childToken is JValue scalar ? scalar.Value ?? JValue.CreateNull() : childToken;
             }
+
+            object? GetJArrayValue(object value, string key, bool ignoreCase)
+            {
+                _ = ignoreCase;
+                var jArr = (JArray)value;
+                switch (key.ToLowerInvariant())
```
Comment (yufeih) [inline src/docfx/template/MustacheTemplate.cs:78]: > perf-tip: do not allocate
```suggestion
                if (string.Equals(key, "length", StringComparer.OrdinalIgnoreCase))
```

File: src/docfx/template/MustacheTemplate.cs [hunk 4]
```diff
@@ -64,6 +70,25 @@ private static RendererSettingsBuilder UseJson(RendererSettingsBuilder settings,
...
+                {
+                    case "length":
+                        return jArr.Count.ToString();
+                    default:
+                        return int.TryParse(key, out var index) && index > -1 && jArr.Count > index
+                            ? jArr[index] is JValue scalar
+                                ? scalar.Value ?? JValue.CreateNull()
+                                : jArr[index]
+                            : JValue.CreateNull();
```
Comment (yufeih) [inline src/docfx/template/MustacheTemplate.cs:87]: Will `null` work? The function should return `JObject`, `JArray` or `primitive type`, `JValue.CreateNull()` is not a primitive type.
```suggestion
                            : null;
```

Comment (mingwli) [reply src/docfx/template/MustacheTemplate.cs:87]: Let's take a try

**Comment by mingwli** [comment]
> ⚠[DevSandbox](https://github.com/OPS-E2E-PPE/DevSandbox)(00:00:23.4002998, 0 more diff)

caused by the {{array.0}} syntax.
```
		{{#enumValues.0}}
			<p>{{__global.enumInSubtitle}} {{__global.values}}:</p>
			<ul class="type-list">{{#enumValues}}<li>{{.}}</li>{{/enumValues}}</ul>
		{{/enumValues.0}}
```
https://ceapex.visualstudio.com/Engineering/_git/docs-ui?path=%2Fpackages%2Fdocfx-templates%2Fpartials%2Fopenapi%2Ftype.tmpl.partial

File: src/docfx/template/MustacheTemplate.cs (hunk 4 above)
Comment (mingwli) [reply src/docfx/template/MustacheTemplate.cs:87]: I didn't touch it due to this PR
https://github.com/dotnet/docfx/pull/4932/files#diff-e57f7b911be9c3fec40a3f7ccc3f17b4R62

File: src/docfx/template/MustacheTemplate.cs [hunk 5]
```diff
@@ -44,26 +44,30 @@ private static RendererSettingsBuilder UseJson(RendererSettingsBuilder settings,
...
-
                 var token = (JObject)value;
                 var childToken = token.GetValue(key, ignoreCase ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal);
 
                 return childToken is JValue scalar ? scalar.Value ?? JValue.CreateNull() : childToken;
             }
+
+            object? GetJValueValue(object value, string key, bool ignoreCase)
+            {
+                _ = ignoreCase;
```
Comment (yufeih) [inline src/docfx/template/MustacheTemplate.cs:68]:
```suggestion
               ~~~ _ = ignoreCase;~~~
```

**Pull request merged**